.gitignore

```diff
@@ -1,2 +1,17 @@
-/target
-/Cargo.lock
+# Generated by Cargo
+# will have compiled files and executables
+debug/
+target/
+
+# Remove Cargo.lock from gitignore if creating an executable, leave it for libraries
+# More information here https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html
+Cargo.lock
+
+# These are backup files generated by rustfmt
+**/*.rs.bk
+
+# MSVC Windows builds of rustc generate these, which store debugging information
+*.pdb
+
+# JetBrains IDEs
+.idea
```

Comment:
Questions on these additions:

1. On my setup, `debug` (as well as `release`) is under `target`. Is that not always the case?
2. I've not encountered `.rs.bk`, does this happen if `rustfmt` misbehaves?
3. Are `*.pdb` files not generated under `target`? Seems to me that all build files should live there.
4. Not a question, but gotta love IDE dotfiles 😅

Reply:
Other than tacking on `.idea`, I plucked this straight from https://github.com/github/gitignore. It gave me a clean repo (Cargo.toml wasn't being ignored before, probably because of the nested projects.)

Reply:
Until we know we need them, let's keep `.gitignore` as small as possible. Otherwise it's hard to know by looking at it what we really need and don't. I can do that before merging, unless you know you need some of the other rules?

Reply:
I think we can safely get rid of `debug` and `*.pdb` since at least in our use case they should end up under `target`. (I'm not sure if there's ever a situation where `debug` could end up elsewhere, but I haven't seen it.) As for `**/*.rs.bk` I imagine you'd only see those if either `rustfmt` or your system in general crashed before `rustfmt` was done flushing the new version of the file to disk (i.e. the process is probably read file -> back up file -> write new file -> delete backup.)

Reply:
Related: https://github.com/rust-lang/rustfmt/issues/1114#issuecomment-341641207